File: libgomp/testsuite/libgomp.oacc-c-c++-common/collapse-2.c
/* { dg-do run } */

#include <stdlib.h>

int
main (void)
{
  int i, j, k, l = 0, f = 0, x = 0, l2 = 0;
  int m1 = 4, m2 = -5, m3 = 17;

#pragma acc parallel
  #pragma acc loop seq collapse(3) reduction(+:l)
    for (i = -2; i < m1; i++)
      for (j = m2; j < -2; j++)
	{
	  for (k = 13; k < m3; k++)
	    {
              if ((i + 2) * 12 + (j + 5) * 4 + (k - 13) !=  9 + f++)
		l++;
	    }
	}

  /*  Test loop with > condition.  */
#pragma acc parallel
  #pragma acc loop seq collapse(3) reduction(+:l2)
    for (i = -2; i < m1; i++)
      for (j = -3; j > (m2 - 1); j--)
	{
	  for (k = 13; k < m3; k++)
	    {
	      if ((i + 2) * 12 + (j + 5) * 4 + (k - 13) !=  9 + f++)
		l2++;
	    }
	}

    for (i = -2; i < m1; i++)
      for (j = m2; j < -2; j++)
	{
	  for (k = 13; k < m3; k++)
	    {
              if ((i + 2) * 12 + (j + 5) * 4 + (k - 13) !=  9 + f++)
		x++;
	    }
	}

  if (l != x || l2 != x)
    abort ();

  return 0;
}

File: libgomp/testsuite/libgomp.oacc-fortran/collapse-2.f90
! { dg-do run }
! { dg-options "-std=legacy" }

program collapse2
  integer :: i, j, k, a(1:3, 4:6, 5:7)
  logical :: l
  l = .false.
  a(:, :, :) = 0
  !$acc parallel
  !$acc loop collapse(4 - 1)
    do 164 i = 1, 3
      do 164 j = 4, 6
        do 164 k = 5, 7
          a(i, j, k) = i + j + k
164      end do
  !$acc end parallel

  !$acc parallel
  !$acc loop collapse(2) reduction(.or.:l)
firstdo: do i = 1, 3
      do j = 4, 6
        do k = 5, 7
          if (a(i, j, k) .ne. (i + j + k)) l = .true.
        end do
      end do
    end do firstdo
  !$acc end parallel
  if (l) STOP 1
end program collapse2
